### DISCUSSION TITLE
Some docker.io HEAD requests have URL appended w/ "@sha256:XXX", result in 404

### DISCUSSION THREAD

**Discussion by MaleNurse** [containrrr/watchtower#860] [Q&A] [answered]
I've noticed that, for some reason, certain images from docker.io have the hash appended to the end of the HEAD request URL, e.g.:

```
time="2021-03-24T09:29:59-07:00" level=debug msg="Doing a HEAD request to fetch a digest" url="https://index.docker.io/v2/containrrr/watchtower/manifests/latest@sha256:ff3ba4f421801d07754150aee4c03fdde26a0a9783d66021e81c7097d6842f25"
```

Whenever this happens, the registry responds with a 404 error:

```
time="2021-03-24T09:30:00-07:00" level=warning msg="Could not do a head request, falling back to regular pull." container=/watchtower_NAS_watchtower-local.1.qzidpy5j299jrks880hcm4u8x image="containrrr/watchtower:latest@sha256:ff3ba4f421801d07754150aee4c03fdde26a0a9783d66021e81c7097d6842f25"

time="2021-03-24T09:30:00-07:00" level=debug msg="Reason: registry responded to head request with &{404 Not Found 404 HTTP/1.1 1 1 map[Content-Length:[19] Content-Type:[text/plain; charset=utf-8] Date:[Wed, 24 Mar 2021 16:30:00 GMT] Docker-Distribution-Api-Version:[registry/2.0] Strict-Transport-Security:[max-age=31536000] X-Content-Type-Options:[nosniff]] {} 19 [] false false map[] 0xc0005a4100 0xc00048e2c0}"
```

Others don't have this problem and work fine:

```
time="2021-03-24T09:30:02-07:00" level=debug msg="Parsing image ref" host=index.docker.io image=mvance/unbound-rpi normalized="docker.io/mvance/unbound-rpi:latest" tag=latest

time="2021-03-24T09:30:02-07:00" level=debug msg="Doing a HEAD request to fetch a digest" url="https://index.docker.io/v2/mvance/unbound-rpi/manifests/latest"

time="2021-03-24T09:30:03-07:00" level=debug msg="Found a remote digest to compare with" remote="sha256:02f4afe40b6a1b1c18761f830633ebf1f2866ed58d30d79fed636f7f89277a72"

time="2021-03-24T09:30:03-07:00" level=debug msg=Comparing local="sha256:02f4afe40b6a1b1c18761f830633ebf1f2866ed58d30d79fed636f7f89277a72" remote="sha256:02f4afe40b6a1b1c18761f830633ebf1f2866ed58d30d79fed636f7f89277a72"

time="2021-03-24T09:30:03-07:00" level=debug msg="Found a match"

time="2021-03-24T09:30:03-07:00" level=debug msg="No pull needed. Skipping image."
```

Looking at the compose files, I can't see any reason why some do this and others don't. I'm using Portainer to deploy these stacks, if that matters at all, and here's the full log from the last run: 
[qzidpy5j299jrks880hcm4u8x_logs.txt](https://github.com/containrrr/watchtower/files/6199155/qzidpy5j299jrks880hcm4u8x_logs.txt)

Is there something I should be looking at to tell me if this is my setup and not watchtower? Should I create an issue about this? Please let me know if there's any more info that might be of use.

**Comment by simonaronsson** [comment]
I think this was solved in `v1.2.1`. Would you be able to verify, @MaleNurse?

**Reply by MaleNurse** [reply]
Sorry for the delay @simskij 

Unfortunately, this still seems to be happening for me. Here's the latest pull of the logs. Let me know if you'd like any more info

[_Watchtower_watchtower-local.1.x9vm1vgjcohyyz2hyd1fuer8a_logs.txt](https://github.com/containrrr/watchtower/files/6371108/_Watchtower_watchtower-local.1.x9vm1vgjcohyyz2hyd1fuer8a_logs.txt)

**Reply by simonaronsson** [reply]
No worries at all! Thanks for getting back! 

Can you `docker inspect` this container and see whether the hash is present there as well? If that's the case, you need to recreate this container without the hash. I'm not a 100% sure why this happens, but it seems related to portainers deployment model as the other issues I've seen on this also have been for portainer users (if my memory serves me right).

If not, we'll have to continue investigating.

**Reply by piksel** [reply]
I think this was stated before, but asking the question "Does this image that is pinned to this exact version have a newer version available?" doesn't really make sense.
You could of course just remove it while checking, but that would potentially break things for others which pinned it for a good reason...

**Reply by simonaronsson** [reply]
> I think this was stated before, but asking the question "Does this image that is pinned to this exact version have a newer version available?" doesn't really make sense.
> You could of course just remove it while checking, but that would potentially break things for others which pinned it for a good reason...

I have a vague recollection of us adding skips for these images at some point, but maybe my memory deceives me. In any case, if I interpret you correctly, @MaleNurse, you didn't actually put the SHA there yourself intentionally, right?

**Comment by MaleNurse** [comment]
Yessir, it's mostly via compose files in portainer, with 3 manually added. I'll get you those inspects, and the compose layouts if you want too, here in a few hours.

Thanks for looking into it!

**Reply by simonaronsson** [reply]
Sounds great! Compose layouts would definitely be helpful as well. Looking forward to it!

**Reply by MaleNurse** [reply]
So here are the container inspection outputs from three of the containers (Watchtower included) where I've seen the head request issue recently. Also in there are all the sanitized compose layouts I'm using:
[watchtower_debugging.zip](https://github.com/containrrr/watchtower/files/6380768/watchtower_debugging.zip)

And here's the latest log pull, just in case:
[_Watchtower_watchtower-local.1.x9vm1vgjcohyyz2hyd1fuer8a_logs.txt](https://github.com/containrrr/watchtower/files/6380775/_Watchtower_watchtower-local.1.x9vm1vgjcohyyz2hyd1fuer8a_logs.txt)

If there's anything else that might be helpful, please let me know. Thanks again

**Reply by simonaronsson** [reply]
Just hit me: are you running docker swarm? @fgierlinger had a similar issue in https://github.com/containrrr/watchtower/issues/752, but we never got any more information on whether they were able to solve it or not.

**Reply by MaleNurse** [reply]
Yeah, actually they are in a swarm. I'm only using that mode for some ancillary functions, so I'll do some reconfiguring to kill it and get back to you on if it makes a difference.

**Reply by MaleNurse** [reply]
@simskij OK, could have sworn I'd gotten back to you guys about this but, apparently, exams are really getting to me. Sorry about that. 
Anyway, leaving the swarm totally worked. No more 404's or other causes for failed HEAD requests and everything is updating nicely. If there's anything else you'd like me to test regarding this, let me know, I'd be glad to help. Otherwise, I'll just keep an ear pricked for a future workaround. Thanks again for the help!

**Reply by mdkeil** [reply]
Hello,
I have exactly the same problem, that some images from docker.io have the hash appended to the end of the HEAD request which ends in "error 404".

Some of them are in swarm mode.. I use portainer to managed all stacks. The "docker-compose"-stacks get updated with no problems.

**Reply by piksel** [reply]
Yeah. As previously stated, we shouldn't attempt to pull images containing the hashes, but it's not obvious that  we should just strip them either. It's even possible (probable?) that doing so would break the swarm.

**Reply by Enissay** [reply]
And, what would be the solution ?
As all my images on my personal machine are from docker.io, the daily email I receive is just full of 404 errors:

`Could not do a head request for "portainer/portainer-ce:latest@sha256:76ff22486bcd3713631b5f317efcb69e707a122fe96ffcc0589cf2d3e8e6b890", falling back to regular pull.
Reason: registry responded to head request with "404 Not Found", auth: "not present"
...`

Or, is there any workaround to make this work for people using portainer ? @piksel

**Reply by piksel** [reply]
You can pass `--warn-on-head-failure never` to avoid the messages, but it would still not update them. It's not portainer per se that is the problem, but rather how docker swarm works.

**Reply by Enissay** [reply]
Well, for sure the goal is not to simply hide them but to make watchtower do its what was meant to.

I am not sure exactly how watchtower works, but the workflow in this case should be as follows:

- check if a new version is available OR simply do a pull of `image:tag` (without hash and I guess tag should be either missing OR latest) and it will be updated accordingly (or not if no new version available)
- If a new version is downloaded, stop the concerned container, else do nothing
- Starting it is not needed since the swarm will do using the new downloaded image (or at least how I configured it)

That's basically how I do update manually (more or less)

**Comment by kb1ibt** [comment, marked as answer]
Reviving the old thread, I am also using swarm, but not using Portainer to deploy the stacks. I am just using `docker stack deploy` and in my docker-compose.yml or docker-stack.yml files I am only supplying `image: traefik:latest`, not pinning it to a specific hash. This is also causing watchtower to generate the 404s because it is searching for the pinned version because in the `docker inspect $(docker ps -aq) | jq '.[].Config.Image'` it is listing the @sha256: 
Even trying to put 
```version: "3"
services:
  watchtower:
    image: containrrr/watchtower
    volumes:
      - /var/run/docker.sock:/var/run/docker.sock
    command: --interval 300
    deploy:
      labels:
        "com.centurylinklabs.watchtower.image": "containrrr/watchtower"
```
doesn't solve the issue because it seems like the label is only kept on the service and is not passed on the container

**Reply by piksel** [reply]
Swarm is pinning the images to sha256 hashes, so watchtower doesn't work with swarm. It's not a user configuration, but rather how swarm interacts with docker.

**Reply by FlyingPersian** [reply]
> Swarm is pinning the images to sha256 hashes, so watchtower doesn't work with swarm. It's not a user configuration, but rather how swarm interacts with docker.

Does this mean that watchtower doesn't work with swarm? I'm having the same issue. I'm running watchtower with `--mode global` on all my nodes, but I also get the 404 error. I just want to use watchtower to notify me if there is an update without updating though.

**Reply by simonaronsson** [reply]
That is what it means. Sorry
